internal/deviations/deviations.go

```diff
@@ -99,4 +108,21 @@ var (
 	StaticProtocolName = flag.String("deviation_static_protocol_name", "DEFAULT", "The name used for the static routing protocol.  The default name in OpenConfig is \"DEFAULT\" but some devices use other names.")
 
 	DeprecatedVlanID = flag.Bool("deviation_deprecated_vlan_id", false, "Device requires using the deprecated openconfig-vlan:vlan/config/vlan-id or openconfig-vlan:vlan/state/vlan-id leaves.")
+
+	AddSubIntfToNetInst = flag.Bool("deviation_adding_subintf_to_netowork_instance", true,
+	"Vendor requiring explicitly attaching sub-interface to network-instance, this deviation should be marked true. Default expectation is sub-interface is part of default network-instance implicitly.")
```

Comment:
```suggestion
	"If a device requires explicit attachment of a sub-interface to a network-instance, this deviation should be marked true. The default expectation is that a sub-interface is part of default network-instance implicitly.")
```

Reply:
I'd suggest that we file a bug in openconfig/public to make sure that a clarfication is made in the model that clearly states that this default association exists.

Reply:
The justification is still unclear to me. It says that the device requires explicitly putting a subinterface into **a** network instance, and the added helper puts the subinterface into an arbitrary (possibly non-default) network instance. But the updated static ARP test puts it into **the default** network instance, which is not required either by OpenConfig spec nor supported by the wording of this deviation.

What happens if the test only configures the subinterface but does not put in into any network instance?

Reply:
This is exactly the problem. The assumption that is made in the schema and tests is that an interface/subinterface is by default assigned to the `DEFAULT-INSTANCE` network instance, which we've now said should be called `DEFAULT`. 

The spec doesn't explicitly say this but it *is* expected, as the tests today concrete this by not assigning interfaces to any network instance. 

It is important to note that the models allow for this explicit assignment (explicitly specifying the default is fine in all cases).

The implementation that is being discussed here does not currently make this default assignment - so we need a way to be able to make it explicitly. In practice, this is a more explicit and likely more compatible config. i.e., if one knows all the interfaces one could make this assignment in all cases and it wouldn't change the behaviour of the device.

For non-default network instances, then this configuration _is_ required explicitly because the implicit default is not the case. So it is not clear that this can be assigned to an arbitrary network-instance. A change to the deviation could be that it only assigns to the default instance - this will need to be dependent upon the default instance name deviation.

Another discussion we could have here is whether we just make this assignment explicit in all tests without the deviation. It seems pragmatic to introduce the deviation to say that this is a "more compatible config, that we don't know whether it's supported in all implementations" and then come back and determine what the preferred approach is going forward after clarifying the specification.

Let's meet if we need to discuss this further.

Reply:
Note that the flag is "deviation_adding_subintf_to_**netowork**_instance". you probably want it to be "deviation_adding_subintf_to_**network**_instance"

Reply:
What is the process for moving an interface/subinterface from one network instance to another? I'm asking because device behavior can differ.

Currently, because the interface/sub does not belong to any specific NI, tests could put them into a specific NI with a single `Replace`. But if we put them into `DEFAULT` explicitly, some devices might require the interface/sub to also be removed explicitly before they can be added to another NI.

If we want tests to consistently put interface/sub into NI, we need to allocate time to update the test plans and tests, and add this requirement to the style guide so new tests will follow it.

My inclination is to just keep the test coverage minimal. If a test does not need NI, then leave it alone. We already have VRF tests (RT-3.1, RT-3.2) that put interface/sub into NI explicitly.

Reply:
> Hi @robshakir, Thanks for the feedback and I'm glad the conversation is started on this subject.
> 
> One additional point that I would like to bring forward is a case where we might assign an interface to a network instance and move it to a different one - (I've not done a deep dive of all test cases to see if that might happen).
> 
> In certain vendors, you might have to explicitly remove the association from it's original Network instance to successfully assign it to the new desired Network Instance.

+1 to the second point

Reply:
The assumption that unreferenced subintefaces always belong to DEFAULT feels like a slippery slope. What other OC modeled objects that support configurable net-instance association should follow the same assumption of belonging to DEFAULT without any explicit binding? Explicit references to DEFAULT ensure that different implementer assumptions about what DEFAULT owns by default do not affect the test result.

Also worth noting.. In many implementations there are checks done to ensure that subinterface references are within scope. For example, it would be common for an implementation to check that all the IS-IS interfaces of an IS-IS instance running in the DEFAULT net-instance actually belong to the DEFAULT network-instance. I believe there is some loss of clarity, for the implementer and operator, if the set of eligible subinterfaces in this case is the superset of unbound subinterfaces and subinterfaces explicitly bound to the DEFAULT network-instance.